plugin/sampling/strategystore/adaptive/processor.go

```diff
@@ -244,53 +214,46 @@ func addJitter(jitterAmount time.Duration) time.Duration {
-				interval:   endTime.Sub(startTime),
-				endTime:    endTime,
-			})
-			lastCheckedTime = endTime
-			// Load the latest throughput so that if this host ever becomes leader, it
-			// has the throughput ready in memory. However, only run the actual calculations
-			// if this host becomes leader.
-			// TODO fill the throughput buffer only when we're leader
-			if p.isLeader() {
-				startTime := time.Now()
-				probabilities, qps := p.calculateProbabilitiesAndQPS()
-				p.Lock()
-				p.probabilities = probabilities
-				p.qps = qps
-				p.Unlock()
-				// NB: This has the potential of running into a race condition if the CalculationInterval
-				// is set to an extremely low value. The worst case scenario is that probabilities is calculated
-				// and swapped more than once before generateStrategyResponses() and saveProbabilities() are called.
-				// This will result in one or more batches of probabilities not being saved which is completely
-				// fine. This race condition should not ever occur anyway since the calculation interval will
-				// be way longer than the time to run the calculations.
-				p.generateStrategyResponses()
-				p.calculateProbabilitiesLatency.Record(time.Since(startTime))
-				p.runBackground(p.saveProbabilitiesAndQPS)
-			}
-		case <-p.shutdown:
-			return
-		}
+func (p *Processor) runCalculation() {
+	endTime := time.Now().Add(p.Delay * -1)
+	startTime := p.lastCheckedTime
+	throughput, err := p.storage.GetThroughput(startTime, endTime)
+	if err != nil {
+		p.logger.Error(getThroughputErrMsg, zap.Error(err))
+		return
+	}
+	aggregatedThroughput := p.aggregateThroughput(throughput)
+	p.prependThroughputBucket(&throughputBucket{
+		throughput: aggregatedThroughput,
+		interval:   endTime.Sub(startTime),
+		endTime:    endTime,
+	})
+	p.lastCheckedTime = endTime
+	// Load the latest throughput so that if this host ever becomes leader, it
+	// has the throughput ready in memory. However, only run the actual calculations
+	// if this host becomes leader.
+	// TODO fill the throughput buffer only when we're leader
+	if p.isLeader() {
+		startTime := time.Now()
+		probabilities, qps := p.calculateProbabilitiesAndQPS()
+		p.Lock()
+		p.probabilities = probabilities
+		p.qps = qps
+		p.Unlock()
+		// NB: This has the potential of running into a race condition if the CalculationInterval
+		// is set to an extremely low value. The worst case scenario is that probabilities is calculated
+		// and swapped more than once before generateStrategyResponses() and saveProbabilities() are called.
+		// This will result in one or more batches of probabilities not being saved which is completely
+		// fine. This race condition should not ever occur anyway since the calculation interval will
+		// be way longer than the time to run the calculations.
+
+		p.calculateProbabilitiesLatency.Record(time.Since(startTime))
+
+		p.bgFinished.Add(1)
+		go func() {
+			p.saveProbabilitiesAndQPS()
+			p.bgFinished.Done()
+		}()
```

Comment:
why do we need to do this in the background when the parent function is already invoked in a loop?

Reply:
Running in the background allows the loop in runAggregationLoop to continue iterating without being blocked by the saving operation.

Reply:
but it is one logical operation, if you break them apart you introduce race conditions

Reply:
Yes, will remove it.